src/coreclr/debug/ee/controller.cpp

```diff
@@ -7553,6 +7638,31 @@ void DebuggerStepper::TriggerUnwind(Thread *thread,
     m_reason = unwindReason;
 }
 
+void DebuggerStepper::TriggerMulticastDelegate(BYTE* pDel, INT32 delegateCount)
+{
+    TraceDestination trace;
+    FramePointer fp = LEAF_MOST_FRAME;
+
+    int totalDelegateCount = (int)*(size_t*)((BYTE*)pDel + DelegateObject::GetOffsetOfInvocationCount());
+    if (delegateCount == totalDelegateCount)
```

Comment:
> If we change the generated IL in the MulticastILStub not to call the helper after the last delegate invocation then we'll never get here when delegateCount==totalDelegateCount 

This would simplify the code emitted for the multicast delegate helper nicely. https://github.com/dotnet/runtime/blob/81efcade9e543a2c03457b3bc87a2bc76b36057e/src/coreclr/vm/comdelegate.cpp#L2146-L2206 can be:

```
        //Label_nextDelegate:
        pCode->EmitLabel(nextDelegate);

#ifdef DEBUGGING_SUPPORTED
        ILCodeLabel *invokeTraceHelper = pCode->NewCodeLabel();
        ILCodeLabel *debuggerCheckEnd = pCode->NewCodeLabel();

        // Call MulticastDebuggerTraceHelper only if any debugger is attached
        pCode->EmitLDC((DWORD_PTR)&g_CORDebuggerControlFlags);
        pCode->EmitCONV_I();
        pCode->EmitLDIND_I4();

        // (g_CORDebuggerControlFlags & DBCF_ATTACHED) != 0
        pCode->EmitLDC(DBCF_ATTACHED);
        pCode->EmitAND();
        pCode->EmitBRTRUE(invokeTraceHelper);

        pCode->EmitLabel(debuggerCheckEnd);
#endif // DEBUGGING_SUPPORTED

        // Load next delegate from array using LoopCounter as index
        pCode->EmitLoadThis();
        pCode->EmitLDFLD(pCode->GetToken(CoreLibBinder::GetField(FIELD__MULTICAST_DELEGATE__INVOCATION_LIST)));
        pCode->EmitLDLOC(dwLoopCounterNum);
        pCode->EmitLDELEM_REF();

        // Load the arguments
        for (UINT paramCount = 0; paramCount < sig.NumFixedArgs(); paramCount++)
            pCode->EmitLDARG(paramCount);

        // call the delegate
        pCode->EmitCALL(pCode->GetToken(pMD), sig.NumFixedArgs(), fReturnVal);

        // Save return value.
        if (fReturnVal)
            pCode->EmitSTLOC(dwReturnValNum);

        // increment counter
        pCode->EmitLDLOC(dwLoopCounterNum);
        pCode->EmitLDC(1);
        pCode->EmitADD();
        pCode->EmitSTLOC(dwLoopCounterNum);

        // compare LoopCounter with InvocationCount. If less then branch to nextDelegate
        pCode->EmitLDLOC(dwLoopCounterNum);
        pCode->EmitLoadThis();
        pCode->EmitLDFLD(pCode->GetToken(CoreLibBinder::GetField(FIELD__MULTICAST_DELEGATE__INVOCATION_COUNT)));
        pCode->EmitBLT(nextDelegate);

        // load the return value. return value from the last delegate call is returned
        if (fReturnVal)
            pCode->EmitLDLOC(dwReturnValNum);

        // return
        pCode->EmitRET();
```